javarosa/src/main/java/org/javarosa/xpath/expr/FunctionUtils.java

```diff
@@ -0,0 +1,447 @@
+            Double n = toInt(o);
+
+            if (n.isNaN()) {
+                return n;
+            }
+
+            if (n.isInfinite() || n > Integer.MAX_VALUE || n < Integer.MIN_VALUE) {
+                throw new XPathTypeMismatchException("converting out-of-range value to date");
+            }
+
+            return DateUtils.dateAdd(DateUtils.getDate(1970, 1, 1), n.intValue());
+        } else if (o instanceof String) {
+            String s = (String)o;
+
+            if (s.length() == 0) {
+                return s;
+            }
+
+            Date d = DateUtils.parseDateTime(s);
+            if (d == null) {
+                throw new XPathTypeMismatchException("converting string " + s + " to date");
+            } else {
+                return d;
+            }
+        } else if (o instanceof Date) {
+            return DateUtils.roundDate((Date)o);
+        } else {
+            String type = o == null ? "null" : o.getClass().getName();
+            throw new XPathTypeMismatchException("converting unexpected type " + type + " to date");
+        }
+    }
+
+    protected static Date expandDateSafe(Object dateObject) {
+        if (!(dateObject instanceof Date)) {
+            // try to expand this out of a nodeset
+            dateObject = toDate(dateObject);
+        }
+        if (dateObject instanceof Date) {
+            return (Date)dateObject;
+        } else {
+            return null;
+        }
+    }
+
+    protected static Object[] subsetArgList(Object[] args, int start) {
+        return subsetArgList(args, start, 1);
+    }
+
+    /**
+     * return a subset of an argument list as a new arguments list
+     *
+     * @param start index to start at
+     * @param skip  sub-list will contain every nth argument, where n == skip (default: 1)
+     */
+    protected static Object[] subsetArgList(Object[] args, int start, int skip) {
+        if (start > args.length || skip < 1) {
+            throw new RuntimeException("error in subsetting arglist");
+        }
+
+        Object[] subargs = new Object[(int)MathUtils.divLongNotSuck(args.length - start - 1, skip) + 1];
+        for (int i = start, j = 0; i < args.length; i += skip, j++) {
+            subargs[j] = args[i];
+        }
+
+        return subargs;
+    }
+
+    public static Object unpack(Object o) {
+        if (o instanceof XPathNodeset) {
+            return ((XPathNodeset)o).unpack();
+        } else {
+            return o;
+        }
+    }
+
+    /**
+     * Perform toUpperCase or toLowerCase on given object.
+     */
+    protected static String normalizeCase(Object o, boolean toUpper) {
+        String s = toString(o);
+        if (toUpper) {
+            return s.toUpperCase();
+        }
+        return s.toLowerCase();
+    }
+
+    /**
+     * Get list of base xpath functions
+     *
+     * (Used in formplayer for function auto-completion)
+     */
+    @SuppressWarnings("unused")
+    public static List<String> xPathFuncList() {
```

Comment:
one thing for vellum, since users don't have to use the latest commcare version (unlike formplayer) some of the functions may not be available for the version they're using. having versions attached to the functions would be useful, but probably not a huge priority